tests/drivers/hwinfo/api/src/main.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <hwinfo.h>
+#include <ztest.h>
+#include <strings.h>
+
+/*
+ * @addtogroup t_hwinfo_get_device_id_api
+ * @{
+ * @defgroup t_hwinfo_get_device_id test_hwinfo_get_device_id
+ * @brief TestPurpose: verify device id get works
+ * @details
+ * - Test Steps
+ *   -# Read the ID
+ *   -# Check if to many bytes are written to the buffer
+ *   -# Check if UID is plausible
+ * - Expected Results
+ *   -# Device uid with correct length should be written to the buffer.
+ * @}
+ */
+
+#define BUFFER_LENGTH 17
+#define BUFFER_CANARY 0xFF
+
+#if defined(CONFIG_DEVICE_ID_STM32) || defined(CONFIG_DEVICE_ID_NRF)
+
+/*
+ * Function invokes the get_entropy callback in driver
+ * to get the random data and fill to passed buffer
+ */
+static int get_device_id(void)
+{
+	struct device *dev;
+	u8_t buffer_1[BUFFER_LENGTH];
+	u8_t buffer_2[BUFFER_LENGTH];
+	size_t length_read_1, length_read_2;
+	int i;
+
+	dev = device_get_binding(CONFIG_DEVICE_ID_NAME);
+	if (!dev) {
+		TC_PRINT("error: no device id device\n");
+		return TC_FAIL;
+	}
+
+	TC_PRINT("devic id device is %p, name is %s\n",
+		 dev, dev->config->name);
+
+	memset(buffer_1, BUFFER_CANARY, sizeof(buffer_1));
+
+	length_read_1 = hwinfo_get_device_id(dev, buffer_1, BUFFER_LENGTH - 1);
+	if (length_read_1 < 0) {
+		TC_PRINT("Fail. Returned %d\n", length_read_1);
+		return TC_FAIL;
+	}
+	if (length_read_1 == 0) {
+		TC_PRINT("Fail. Length is 0");
+		return TC_FAIL;
+	}
+	if (length_read_1 > BUFFER_LENGTH - 1) {
+		TC_PRINT("Fail. Length not adhered\n");
+		return TC_FAIL;
+	}
+
+	if (buffer_1[length_read_1] != BUFFER_CANARY) {
+		TC_PRINT("Fail. Too many bytes are written\n");
+		return TC_FAIL;
+	}
+
+	memcpy(buffer_2, buffer_1, length_read_1);
+
```

Comment:
This should be `memcpy(buffer_2, buffer_1, BUFFER_LENGTH)` or the bytes past `length_read_1` will be uninitialized.

Reply:
The bytes after length_read_1 are not checked. Do you want me to initialize it anyway?
How did you get this running on nRF5? I tried with the nRF52 dongel but it does not compile because of my userspace problem.

Reply:
I mis-read the code: thought you were leaving `buffer_1` alone, and reading into `buffer_2` for the second pass.  So this should be fine.

I know nothing about userspace; I hacked `hello_world` like this:
```
diff --git a/samples/hello_world/prj.conf b/samples/hello_world/prj.conf
index b2a4ba5910..c3fa436ef9 100644
--- a/samples/hello_world/prj.conf
+++ b/samples/hello_world/prj.conf
@@ -1 +1,3 @@
 # nothing here
+CONFIG_DEVICE_ID=y
+CONFIG_STDOUT_CONSOLE=y
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index 2d4a7f0d5f..d53ef17873 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -5,9 +5,26 @@
  */
 
 #include <zephyr.h>
-#include <misc/printk.h>
+#include <hwinfo.h>
+#include <stdio.h>
 
 void main(void)
 {
-       printk("Hello World! %s\n", CONFIG_BOARD);
+       printf("Hello World! %s\n", CONFIG_BOARD);
+
+       struct device *hwp = device_get_binding(CONFIG_DEVICE_ID_NAME);
+       printk("HW device %s at %p\n", CONFIG_DEVICE_ID_NAME, hwp);
+       if (hwp) {
+               u8_t id[16];
+               int rc = hwinfo_get_device_id(hwp, id, sizeof(id));
+               printf("HWID got %d:", rc);
+               if (0 < rc) {
+                       const u8_t * ip = id;
+                       const u8_t * const ipe = ip + rc;
+                       while (ip < ipe) {
+                               printf(" %02x", *ip++);
+                       }
+               }
+               printf("\n");
+       }
 }
```